plugins/destination/mssql/client/migrate.go

```diff
@@ -2,48 +2,190 @@ package client
+	if err != nil {
+		c.logErr(err)
+		return nil, err
+	}
+
+	schemaTables := make(schema.Tables, 0)
+	if err := processRows(rows, func(row *sql.Rows) error {
+		var tableCatalog string
+		var tableName string
+		var tableType string
+		var schemaType string
+
+		if err := row.Scan(&tableCatalog, &tableType, &tableName, &schemaType); err != nil {
 			return err
 		}
+		if tables.Get(tableName) == nil {
+			return nil
+		}
+		schemaTables = append(schemaTables, &schema.Table{Name: tableName})
+		return nil
+	}); err != nil {
+		c.logErr(err)
+		return nil, err
 	}
 
-	return nil
+	for _, table := range schemaTables {
+		columns, err := c.getTableColumns(ctx, table)
+		if err != nil {
+			return nil, err
+		}
+		table.Columns = columns
+		pks, err := c.getTablePK(ctx, table)
+		if err != nil {
+			return nil, err
+		}
+		for _, pk := range pks {
+			if col := table.Columns.Get(pk); col != nil {
+				col.CreationOptions.PrimaryKey = true
+			}
+		}
+	}
+
+	return schemaTables, nil
 }
 
-func (c *Client) migrate(ctx context.Context, table *schema.Table) (err error) {
-	c.logger.Info().Str("table", table.Name).Msg("Migrating table started")
-	defer c.logger.Info().Str("table", table.Name).Msg("Migrating table done")
-	defer func() {
-		if err != nil {
-			c.logErr(err)
+func (c *Client) normalizedTables(tables schema.Tables) schema.Tables {
+	allTables := tables.FlattenTables()
+	var normalized schema.Tables
+	for _, table := range allTables {
+		tableCopy := table.Copy(table.Parent)
+		for i := range tableCopy.Columns {
+			// Since multiple schema types can map to the same MSSQL type we need to normalize them to avoid false positives when detecting schema changes
+			tableCopy.Columns[i].Type = queries.SchemaType(queries.SQLType(table.Columns[i].Type))
 		}
-	}()
+		// If there are no PKs, we use CqID as PK
+		pks := tableCopy.PrimaryKeys()
+		if !c.pkEnabled() || len(pks) == 0 {
+			tableCopy.Columns.Get(schema.CqIDColumn.Name).CreationOptions.PrimaryKey = true
+		}
+
+		normalized = append(normalized, tableCopy)
+	}
+
+	return normalized
+}
 
-	if len(table.Columns) == 0 {
-		c.logger.Info().Str("table", table.Name).Msg("Table with no columns, skipping")
+func (c *Client) nonAutoMigrableTables(tables schema.Tables, schemaTables schema.Tables) ([]string, [][]schema.TableColumnChange) {
+	var result []string
+	var tableChanges [][]schema.TableColumnChange
+	for _, t := range tables {
+		schemaTable := schemaTables.Get(t.Name)
+		if schemaTable == nil {
+			continue
+		}
+		changes := t.GetChanges(schemaTable)
+		if !c.canAutoMigrate(changes) {
+			result = append(result, t.Name)
+			tableChanges = append(tableChanges, changes)
+		}
+	}
+	return result, tableChanges
+}
+
+func (*Client) canAutoMigrate(changes []schema.TableColumnChange) bool {
+	for _, change := range changes {
+		if change.Type == schema.TableColumnChangeTypeAdd && (change.Current.CreationOptions.PrimaryKey || change.Current.CreationOptions.NotNull) {
```

Comment:
might be useful to extract a simple func
```go
func notNull(options schema.ColumnCreationOptions) bool {
	return options.NotNull || options.PrimaryKey
}
```
and use it here and in other places (normalization?)

Reply:
Not sure that adds clarity as it's used only in this function so I don't see a reason to skip to another function to follow the logic.
Also the name `notNull` is not correct it should be something like `isOptionBreaking` or something like that. So a bit hard to explain out of the context of this function